src/doc/trpl/closures.md

```diff
@@ -1,214 +1,510 @@
+Now that we know that closures are traits, we already know how to accept and
+return closures: just like any other trait!
 
-Okay, let's look at the body of `twice`:
+This also means that we can choose static vs dynamic dispatch as well. First,
+let's write a function which takes a closure, calls it, and returns the
+result:
 
-```{rust}
-fn twice<F: Fn(i32) -> i32>(x: i32, f: F) -> i32 {
-  f(x) + f(x)
+```rust
+fn call_with_one<F>(some_closure: F) -> i32
+    where F : Fn(i32) -> i32 {
+
+  some_closure(1)
 }
+
+let answer = call_with_one(|x| x + 2);
+
+assert_eq!(3, answer);
 ```
 
-Since our closure is named `f`, we can call it just like we called our closures
-before, and we pass in our `x` argument to each one, hence the name `twice`.
+We pass our closure, `|x| x + 2`, to `call_with_one`. It just does what it
+suggests: it calls the closure, giving it `1` as an argument.
 
-If you do the math, `(5 * 5) + (5 * 5) == 50`, so that's the output we get.
+Let's examine the signature of `call_with_one` in more depth:
 
-Play around with this concept until you're comfortable with it. Rust's standard
-library uses lots of closures where appropriate, so you'll be using
-this technique a lot.
+```rust
+fn call_with_one<F>(some_closure: F) -> i32
+#    where F : Fn(i32) -> i32 {
+# some_closure(1) }
+```
 
-If we didn't want to give `square` a name, we could just define it inline.
-This example is the same as the previous one:
+We take one parameter, and it has the type `F`. We also return a `i32`. This part
+isn't interesting. The next part is:
 
-```{rust}
-fn twice<F: Fn(i32) -> i32>(x: i32, f: F) -> i32 {
-    f(x) + f(x)
-}
+```rust
+# fn call_with_one<F>(some_closure: F) -> i32
+    where F : Fn(i32) -> i32 {
+# some_closure(1) }
+```
 
-fn main() {
-    twice(5, |x: i32| { x * x }); // evaluates to 50
+Because `Fn` is a trait, we can bound our generic with it. In this case, our closure
+takes a `i32` as an argument and returns an `i32`, and so the generic bound we use
+is `Fn(i32) -> i32`.
+
+There's one other key point here: because we're bounding a generic with a
+trait, this will get monomorphized, and therefore, we'll be doing static
+dispatch into the closure. That's pretty neat. In many langauges, closures are
+inherently heap allocated, and will always involve dynamic dispatch. In Rust,
+we can stack allocate our closure environment, and statically dispatch the call.
+
+Of course, if we want dynamic dispatch, we can get that too. A trait object
+handles this case, as usual:
+
+```rust
+fn call_with_one(some_closure: &Fn(i32) -> i32) -> i32 {
+  some_closure(1)
 }
+
+let answer = call_with_one(&|x| x + 2);
+
+assert_eq!(3, answer);
 ```
 
-A named function's name can be used wherever you'd use a closure. Another
-way of writing the previous example:
+Now we take a trait object, a `&Fn`. And we have to make a reference
+to our closure when we pass it to `call_with_one`, so we use `&||`.
+
+# Returning closures
```

Comment:
what do you think some good reasons are? I know I've done this, but can't come up with a canonical reason off the top of my head